#159341120 add user login

damafeez opened the pull request:
#### What does this PR do?

- add login method to user model
- add login controller
- add login route
- add passing login test
- return user token

#### Description of Task to be completed?
none

#### How should this be manually tested?
`npm test`
Or by sending a POST resquest to /auth/login via post man after the app has been started with `npm start`

#### Any background context you want to provide?
n/a

#### What are the relevant pivotal tracker stories?
[#159341120](https://www.pivotaltracker.com/n/projects/2183381)
Screenshots (if appropriate)
n/a

Questions:
None

damafeez committed:
ft(server): user should be able to login

- add failing user test

[Finishes #159341120]

damafeez committed:
ft(server): user should be able to login

 - complete route /auth/login
- login.spec passing

    [Finishes #159341120]

damafeez committed:
ft(server): user should be able to login

- add login method to - user model
- add login controller
- add login route
- add passing login - test
- return user token

[finished #159341120]

fidelisojeah commented on server/controllers/user.js:
```diff
@@ -21,3 +21,23 @@ export async function signup(request, response) {
+    } = request.body;
+    const returningUser = new User({ username, password });
+    const user = await returningUser.login();
+    response.status(200).header('x-auth-token', user.token).json({
```
send the token back to the user as part of the json object

It's when the user makes request that they send to the server as part of their header

fidelisojeah commented on server/models/User.js:
```diff
@@ -48,10 +59,40 @@ export default class User {
 
+  async login() {
+    try {
+      const authQuery = `SELECT * FROM authentication WHERE username = '${this.username}'`;
```
So instead of performing 2 SQL queries,
You could select from both tables and join

Something like:
```sql
SELECT users.id, authentication.password, authentication.username, users.fullName, users.email FROM authentication INNER JOIN users 
ON users.authId = authentication.authId 
WHERE authentication.username = '${this.username}'
```

I didn't test this though so may need modification

damafeez replied:
Ok, I will.

fidelisojeah commented on server/models/User.js:
```diff
@@ -48,10 +59,40 @@ export default class User {
+    }
+  }
+
+  genToken() {
```
Try using the asynchronous version of the jwt.sign (Add a callback)
```javascript
getToken(){
  return new Promise((resolve,reject) =>{
    jwt.sign({
       exp: (Math.floor(Date.now() / 1000) + (60 * 60)) * 24 * 7,
       data: this.strip(),
     }, process.env.JWT_SECRET, (error, signedToken)=>{
        if (error){
          reject(new Error(error.message));
        }else{
          resolve(signedToken);
        }
  })
})
}
```

Then call the asynchronous version of this function as well

damafeez replied:
Ok, I'll do try this

damafeez commented on server/models/User.js:
```diff
@@ -48,10 +59,40 @@ export default class User {
+        this.id = getUser.rows[0].id;
+        this.email = getUser.rows[0].email;
+        this.fullName = getUser.rows[0].fullName;
+        this.token = this.genToken();
```
Yes, I'm sending the token back.
That is why I did `this.token = this.genToken();`
I send `this` back to the user (after the strip function removes it's password and authId)
@fidelisojeah

damafeez committed:
ft(server): user should be able to login

    - add sendResponse utils for code reusability
    - use simpler async genToken
    [finished #159341120]

phemmz commented on server/models/User.js:
```diff
@@ -48,10 +59,39 @@ export default class User {
+      const getUser = await client.query(authQuery);
+      const user = getUser.rows[0];
+      if (!user) return new Error('user not found');
+      const correctPassword = await bcrypt.compare(this.password, getUser.rows[0].password);
```
Since this variable value will be a boolean, I think you can use a different name for this variable

damafeez committed:
ft(server): user should be able to login

- rename correctPassword to conform with Boolean naming convention

[finishes #159341120]
